File: GCDMOD.cpp
#include<bits/stdc++.h>
using namespace std;
long long int power(long long int a,long long int b,long long int p)
{
	long long int x,y,z,q,r;
	if(b==0)
		return 1;
	if(a>1000000000)
	{
		x=floor(a/1000000000);
		z=a-100000000*x;
		y=(((1000000000000000000%p)*x*x)%p+(((2*1000000000*x)%p)*z)%p+(z*z)%p)%p;
	}
	else
		y=(a*a)%p;
	if(b%2==0)
		return (power(y,b/2,p))%p;
	else
	{
		q=power(y,(b-1)/2,p);
		if(q>1000000)
		{
			x=floor(q/1000000);
			z=q-1000000*x;
			r=((a%p)*((1000000*x)%p))%p+
		}
		return (a*power(y,(b-1)/2,p))%p;
	}
}
long long int gcd(long long int a,long long int b)
{
	if(b==0)
		return a;
	if(a%b==0)
		return b;
	return gcd(b,a%b);
}
int main()
{
	long long int t,a,b,n,i,j,x,y,z,p;
	cin>>t;
	while(t--)
	{
		cin>>a>>b>>n;
		p=a-b;
		if(p==0)
		{
			x=power(a,n,1000000007);
			y=power(b,n,1000000007);
			z=(x+y)%1000000007;
			cout<<z<<"\n";
		}
		else
		{
			x=power(a,n,p);
			y=power(b,n,p);
			z=(x+y)%p;
			cout<<gcd(p,z)%1000000007<<"\n";
		}
	}
	return 0;
} 

File: GCDMOD.java
import java.io.*;
import java.util.*;
import java.math.BigInteger;
class GCDMOD
{
	public static void main(String args[])throws IOException
	{
		int t;
		BigInteger a,b,n,x,y,z,p,pp=new BigInteger("1000000007");
		Scanner sc=new Scanner(System.in);
		t=sc.nextInt();
		while(t-->0)
		{
			a=sc.nextBigInteger();
			b=sc.nextBigInteger();
			n=sc.nextBigInteger();
			p=a.subtract(b);
			if(a.compareTo(b)==0)
			{
				x=power(a,n,pp);
				y=power(b,n,pp);
				z=(x.add(y)).mod(pp);
				System.out.println(z);
			}
			else
			{
				x=power(a,n,p);
				y=power(b,n,p);
				z=(x.add(y)).mod(p);
				System.out.println(gcd(p,z).mod(pp));
			}
		}	


	}
	public static BigInteger power(BigInteger a,BigInteger b,BigInteger p)
	{
		BigInteger y=new BigInteger("2");
		BigInteger z;
		if(b.compareTo(new BigInteger("0"))==0)
		{
			z=new BigInteger("1");
			return z;
		}
		if(b.mod(y).compareTo(new BigInteger("0"))==0)
		{
			z=(power((a.multiply(a)).mod(p),b.divide(y),p)).mod(p);
			return z;
		}
		else
		{
			z=a.multiply(power((a.multiply(a)).mod(p),(b.subtract(new BigInteger("1"))).divide(y),p)).mod(p);
			return z;
		}
	}
	public static BigInteger gcd(BigInteger a,BigInteger b)
	{
		if(b.compareTo(new BigInteger("0"))==0)
			return a;
		if(a.mod(b).compareTo(new BigInteger("0"))==0)
			return b;
		return gcd(b,a.mod(b));
	}
}